### PR TITLE
Overall bounding box service

### PR DESCRIPTION
**Pull request by 104H**

### CODE REVIEW

File: examples/python/gRPC/meta/gRPC_fb_getOverallBoundingBox.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    + " , "
+    + str(response.PointMax().Z())
+)
+
+responseBuf = stub.GetOverallTimeInterval(bytes(buf))
+response = TimeInterval.TimeInterval.GetRootAs(responseBuf)
+
+print("Min Time (Seconds, Nanoseconds): " + str(response.TimeMin().Seconds()) + " ," + str(response.TimeMin().Nanos()))
+
+print("Max Time (Seconds, Nanoseconds): " + str(response.TimeMax().Seconds()) + " ," + str(response.TimeMax().Nanos()))
```
Comment (Mark-Niemeyer) [inline examples/python/gRPC/meta/gRPC_fb_getOverallBoundingBox.py:65]: remove or rename file

File: seerep-com/fbs/meta_operations.fbs [hunk 2]
```diff
@@ -3,6 +3,10 @@ include "project_info.fbs";
 include "project_infos.fbs";
 include "projectCreation.fbs";
 include "empty.fbs";
+include "boundingbox.fbs";
+include "datatype.fbs";
```
Comment (Mark-Niemeyer) [inline seerep-com/fbs/meta_operations.fbs:7]: Is this include really needed?

File: seerep-msgs/fbs/uuid_datatype_pair.fbs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
+include "datatype.fbs";
+
+namespace seerep.fb;
+
+table UuidDatatypePair {
+    projectuuid:string;
+    datatypes:Datatype;
```
Comment (Mark-Niemeyer) [inline seerep-msgs/fbs/uuid_datatype_pair.fbs:8]: rename to singular

File: seerep-srv/seerep-core/include/seerep-core/core-dataset.h [hunk 4]
```diff
@@ -103,6 +103,14 @@ class CoreDataset
   void addLabels(const seerep_core_msgs::Datatype& datatype, const std::vector<std::string>& labels,
                  const boost::uuids::uuid& msgUuid);
 
+  /**
+   * @brief Get the Time Bounds object
+   *
+   * @return seerep_core_msgs::AABB
+   */
+  seerep_core_msgs::AabbTime getTimeBounds(std::vector<seerep_core_msgs::Datatype> datatypes);
+  seerep_core_msgs::AABB getSpatialBounds(std::vector<seerep_core_msgs::Datatype> datatypes);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-core/include/seerep-core/core-dataset.h:112]: update code docu

File: seerep-srv/seerep-core/include/seerep-core/core-project.h [hunk 5]
```diff
@@ -132,6 +132,9 @@ class CoreProject
    */
   std::shared_ptr<HighFive::File> getHdf5File();
 
+  seerep_core_msgs::AabbTime getTimeBounds(std::vector<seerep_core_msgs::Datatype> datatypes);
+  seerep_core_msgs::AABB getSpatialBounds(std::vector<seerep_core_msgs::Datatype> datatypes);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-core/include/seerep-core/core-project.h:136]: add code docu

File: seerep-srv/seerep-core/include/seerep-core/core.h [hunk 6]
```diff
@@ -128,6 +128,10 @@ class Core
    */
   void deleteProject(boost::uuids::uuid uuid);
 
+  seerep_core_msgs::AabbTime getOverallTimeInterval(boost::uuids::uuid uuid,
+                                                    std::vector<seerep_core_msgs::Datatype> datatypes);
+  seerep_core_msgs::AABB getOverallBound(boost::uuids::uuid uuid, std::vector<seerep_core_msgs::Datatype> datatypes);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-core/include/seerep-core/core.h:133]: add code docu

File: seerep-srv/seerep-server/src/fb-meta-operations.cpp [hunk 7]
```diff
@@ -144,4 +144,130 @@ grpc::Status FbMetaOperations::DeleteProject(grpc::ServerContext* context,
...
+                                         flatbuffers::grpc::Message<seerep::fb::TimeInterval>* response)
+{
+  (void)context;  // ignore that variable without causing warnings
+  auto requestRoot = request->GetRoot();
+
+  std::string uuid = request->GetRoot()->projectuuid()->str();
+  boost::uuids::string_generator gen;
+  auto uuidFromString = gen(uuid);
+
+  std::vector<seerep_core_msgs::Datatype> dt_vector;
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/fb-meta-operations.cpp:159]: please add a comment why you are using a vector, but only adds one element. Short summary of the problem with the fb enum

File: seerep-srv/seerep-server/src/fb-meta-operations.cpp [hunk 8]
```diff
@@ -144,4 +144,130 @@ grpc::Status FbMetaOperations::DeleteProject(grpc::ServerContext* context,
...
+  else if (casted_datatype == seerep::fb::Datatype_Point)
+  {
+    dt_vector.push_back(seerep_core_msgs::Datatype::Point);
+  }
+  else
+  {
+    dt_vector.push_back(seerep_core_msgs::Datatype::Unknown);
+  }
+
+  seerep_core_msgs::AabbTime timeinterval = seerepCore->getOverallTimeInterval(uuidFromString, dt_vector);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/fb-meta-operations.cpp:179]: All packages in the seerep-server package should catch all errors. Take a look at the other functions and add the try-catch here

File: seerep-srv/seerep-server/src/fb-meta-operations.cpp [hunk 9]
```diff
@@ -144,4 +144,130 @@ grpc::Status FbMetaOperations::DeleteProject(grpc::ServerContext* context,
...
+  else if (casted_datatype == seerep::fb::Datatype_Point)
+  {
+    dt_vector.push_back(seerep_core_msgs::Datatype::Point);
+  }
+  else
+  {
+    dt_vector.push_back(seerep_core_msgs::Datatype::Unknown);
+  }
+
+  seerep_core_msgs::AABB overallBB = seerepCore->getOverallBound(uuidFromString, dt_vector);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/fb-meta-operations.cpp:246]: All packages in the seerep-server package should catch all errors. Take a look at the other functions and add the try-catch here

File: seerep-srv/seerep-server/src/fb-meta-operations.cpp [hunk 10]
```diff
@@ -144,4 +144,130 @@ grpc::Status FbMetaOperations::DeleteProject(grpc::ServerContext* context,
...
+                                        flatbuffers::grpc::Message<seerep::fb::Boundingbox>* response)
+{
+  (void)context;  // ignore that variable without causing warnings
+  auto requestRoot = request->GetRoot();
+
+  std::string uuid = requestRoot->projectuuid()->str();
+  boost::uuids::string_generator gen;
+  auto uuidFromString = gen(uuid);
+
+  std::vector<seerep_core_msgs::Datatype> dt_vector;
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/fb-meta-operations.cpp:226]: please add a comment why you are using a vector, but only adds one element. Short summary of the problem with the fb enum

File: seerep-srv/seerep-core/src/core-dataset.cpp [hunk 11]
```diff
@@ -489,4 +489,88 @@ void CoreDataset::addLabels(const seerep_core_msgs::Datatype& datatype, const st
   }
 }
 
+seerep_core_msgs::AabbTime CoreDataset::getTimeBounds(std::vector<seerep_core_msgs::Datatype> datatypes)
+{
+  seerep_core_msgs::AabbTime overalltime;
+
+  overalltime.max_corner().set<0>(0);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-core/src/core-dataset.cpp:496]: Shouldn't this be std::numeric_limits<int64_t>::min()  ?

File: seerep-srv/seerep-core/src/core-dataset.cpp [hunk 12]
```diff
@@ -489,4 +489,88 @@ void CoreDataset::addLabels(const seerep_core_msgs::Datatype& datatype, const st
...
+}
+
+seerep_core_msgs::AABB CoreDataset::getSpatialBounds(std::vector<seerep_core_msgs::Datatype> datatypes)
+{
+  seerep_core_msgs::AABB overallbb;
+
+  // set the minimum to zero
+  overallbb.max_corner().set<0>(0);
+  overallbb.max_corner().set<1>(0);
+  overallbb.max_corner().set<2>(0);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-core/src/core-dataset.cpp:526]: Shouldn't this be std::numeric_limits\<float>::min() ?

**Review by Mark-Niemeyer — CHANGES_REQUESTED** [review]

**Comment by Mark-Niemeyer** [comment]
Also add the pb-interface, please.

File: seerep-srv/seerep-server/src/pb-meta-operations.cpp [hunk 13]
```diff
@@ -77,4 +77,133 @@ grpc::Status PbMetaOperations::GetProjects(grpc::ServerContext* context, const g
...
+    std::string uuid = request->projectuuid();
+    boost::uuids::string_generator gen;
+    auto uuidFromString = gen(uuid);
+
+    seerep_core_msgs::AabbTime timeinterval = seerepCore->getOverallTimeInterval(uuidFromString, dt_vector);
+
+    // isolate second and nano second bits from min time
+    uint64_t mintime = timeinterval.min_corner().get<0>();
+    uint32_t min_nanos = (uint32_t)mintime;
+    uint32_t min_seconds = (uint32_t)(mintime >> 32);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/pb-meta-operations.cpp:117]: double check signed/unsigned

from seerep-msgs/core/timestamp.h:
```
  int32_t seconds;
  uint32_t nanos;
```

File: seerep-srv/seerep-server/src/pb-meta-operations.cpp [hunk 14]
```diff
@@ -77,4 +77,133 @@ grpc::Status PbMetaOperations::GetProjects(grpc::ServerContext* context, const g
...
+
+    // isolate second and nano second bits from max time
+    uint64_t maxtime = timeinterval.max_corner().get<0>();
+    uint32_t max_nanos = (uint32_t)maxtime;
+    uint32_t max_seconds = (uint32_t)(maxtime >> 32);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/pb-meta-operations.cpp:122]: double check signed/unsigned

from seerep-msgs/core/timestamp.h:
```
  int32_t seconds;
  uint32_t nanos;
```

File: seerep-srv/seerep-server/src/fb-meta-operations.cpp [hunk 15]
```diff
@@ -144,4 +144,192 @@ grpc::Status FbMetaOperations::DeleteProject(grpc::ServerContext* context,
...
+    std::string uuid = request->GetRoot()->projectuuid()->str();
+    boost::uuids::string_generator gen;
+    auto uuidFromString = gen(uuid);
+
+    seerep_core_msgs::AabbTime timeinterval = seerepCore->getOverallTimeInterval(uuidFromString, dt_vector);
+
+    // isolate second and nano second bits from min time
+    uint64_t mintime = timeinterval.min_corner().get<0>();
+    uint32_t min_nanos = (uint32_t)mintime;
+    uint32_t min_seconds = (uint32_t)(mintime >> 32);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/fb-meta-operations.cpp:193]: double check signed/unsigned

from seerep-msgs/core/timestamp.h:
```
  int32_t seconds;
  uint32_t nanos;
```

File: seerep-srv/seerep-server/src/fb-meta-operations.cpp [hunk 16]
```diff
@@ -144,4 +144,192 @@ grpc::Status FbMetaOperations::DeleteProject(grpc::ServerContext* context,
...
+
+    // isolate second and nano second bits from max time
+    uint64_t maxtime = timeinterval.max_corner().get<0>();
+    uint32_t max_nanos = (uint32_t)maxtime;
+    uint32_t max_seconds = (uint32_t)(maxtime >> 32);
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-server/src/fb-meta-operations.cpp:198]: double check signed/unsigned

from seerep-msgs/core/timestamp.h:
```
  int32_t seconds;
  uint32_t nanos;
```

File: seerep-srv/seerep-core/src/core-dataset.cpp [hunk 17]
```diff
@@ -517,4 +517,88 @@ void CoreDataset::addLabels(const seerep_core_msgs::Datatype& datatype,
...
+}
+
+seerep_core_msgs::AABB CoreDataset::getSpatialBounds(std::vector<seerep_core_msgs::Datatype> datatypes)
+{
+  seerep_core_msgs::AABB overallbb;
+
+  // set the minimum to zero
+  overallbb.max_corner().set<0>(std::numeric_limits<int64_t>::min());
+  overallbb.max_corner().set<1>(std::numeric_limits<int64_t>::min());
+  overallbb.max_corner().set<2>(std::numeric_limits<int64_t>::min());
```
Comment (Mark-Niemeyer) [inline seerep-srv/seerep-core/src/core-dataset.cpp:554]: This should also be `float` and not `int64_t`, shouldn't it?

**Review by Mark-Niemeyer — CHANGES_REQUESTED** [review]

**Comment by 104H** [comment]
Closed due to erroneous rebase

**Pull request closed without merging**